src/yz_solrq_sup.erl

```diff
@@ -20,70 +20,112 @@
 %%% API functions
 %%%===================================================================
 
 start_link() ->
-    start_link(worker_procs(), helper_procs()).
-
-start_link(NumQueues, NumHelpers) ->
-    supervisor:start_link({local, ?MODULE}, ?MODULE, [NumQueues, NumHelpers]).
+    supervisor:start_link({local, ?MODULE}, ?MODULE, []).
 
 
 %% @doc Start the drain FSM, under this supervisor
 -spec start_drain_fsm(proplist()) -> {ok, pid()} | {error, Reason :: term()}.
-start_drain_fsm(CallbackList) ->
+start_drain_fsm(Parameters) ->
+    PartitionToDrain = proplists:get_value(
+        ?DRAIN_PARTITION, Parameters, undefined
+    ),
     supervisor:start_child(
         ?MODULE,
-        {yz_solrq_drain_fsm, {yz_solrq_drain_fsm, start_link, [CallbackList]}, temporary, 5000, worker, []}
+        {PartitionToDrain, {yz_solrq_drain_fsm, start_link, [Parameters]}, temporary, 5000, worker, []}
     ).
-
+-spec start_queue_pair(Index::index_name(), Partition::p()) -> ok.
+start_queue_pair(Index, Partition) ->
+    validate_child_started(
+        supervisor:start_child(?MODULE, queue_pair_spec({Index, Partition}))).
+
+active_queues() ->
+    PairSups = find_pair_supervisors(),
+    PairChildren = lists:flatten([supervisor:which_children(Sup) || Sup <- PairSups]),
+    Workers = [{Index, Partition} || {{worker, Index, Partition}, _Child, _Type, Modules} <- PairChildren, Modules == [yz_solrq_worker]],
+    Workers.
+
+find_pair_supervisors() ->
+    AllChildren = supervisor:which_children(yz_solrq_sup),
+    PairSups = [SupPid || {_IndexPartition, SupPid, _Type, Modules} <- AllChildren, Modules == [yz_solrq_queue_pair_sup]],
+    PairSups.
 
 %%%===================================================================
 %%% Supervisor callbacks
 %%%===================================================================
 
-init([NumQueues, NumHelpers]) ->
-    yz_solrq:set_solrq_worker_tuple(NumQueues),
-    yz_solrq:set_solrq_helper_tuple(NumHelpers),
+init([]) ->
+
     DrainMgrSpec = {yz_solrq_drain_mgr, {yz_solrq_drain_mgr, start_link, []}, permanent, 5000, worker, [yz_drain_mgr]},
-    QueueChildren = [child_spec(worker, Name) ||
-                        Name <- tuple_to_list(yz_solrq:get_solrq_worker_tuple())],
-    HelperChildren = [child_spec(helper, Name) ||
-                        Name <- tuple_to_list(yz_solrq:get_solrq_helper_tuple())],
+
+    QueueChildren = [queue_pair_spec(IndexPartition) ||
+                        IndexPartition <- required_queues()],
     %% Using a one_for_all restart strategy as we write data to a hashed worker,
     %% which then uses a random helper to send data to Solr itself - if we want to
     %% make this one_for_one we will need to do more work monitoring the processes
     %% and responding to crashes more carefully.
-    {ok, {{one_for_all, 10, 10}, [DrainMgrSpec | HelperChildren ++ QueueChildren]}}.
+    {ok, {{one_for_all, 10, 10}, [DrainMgrSpec | QueueChildren]}}.
```

Comment:
Can we remove the above comments and make this a `one_for_one` supervisor? It seems like we should be able to, since each worker now has its own personal helper which is running under a `one_for_all` supervisor.